crates/engine/tree/src/tree/payload_validator.rs

```diff
@@ -522,11 +526,100 @@ where
         // terminate prewarming task with good state output
         handle.terminate_caching(Some(&output.state));
 
+        // Capture parent hash and ancestor overlays for deferred trie input construction.
+        let (parent_hash, overlay_blocks) = ctx
+            .state()
+            .tree_state
+            .blocks_by_hash(block.parent_hash())
+            .unwrap_or_else(|| (block.parent_hash(), Vec::new()));
+
+        // Create a deferred handle to store the sorted trie data.
+        let deferred_trie_data = DeferredTrieData::pending();
+        let deferred_handle_task = deferred_trie_data.clone();
+        let deferred_handle_err = deferred_trie_data.clone();
+        let block_hash = block.hash();
+        let hashed_state_for_trie = hashed_state;
+        let trie_output_for_trie = trie_output;
+        let overlay_blocks_for_trie = overlay_blocks;
+        let deferred_compute_duration =
+            self.metrics.block_validation.deferred_trie_compute_duration.clone();
+
+        // Defer trie sorting and compute trie input to a background task so the validation hot path
+        // can return once the state root is checked. Consumers (DB writes, overlay
+        // providers, proofs) block on the handle only when they actually need the sorted
+        // trie data.
+        let task = move || {
+            let compute_start = Instant::now();
+            let result = std::panic::catch_unwind(AssertUnwindSafe(|| {
+                let mut parent_trie_input = TrieInputSorted::default();
+                let mut blocks_iter = overlay_blocks_for_trie.iter().rev().peekable();
+
+                if let Some(first) = blocks_iter.next() {
+                    let data = first.trie_data();
+                    parent_trie_input.state = data.hashed_state;
+                    parent_trie_input.nodes = data.trie_updates;
+
+                    if blocks_iter.peek().is_some() {
+                        let state_mut = Arc::make_mut(&mut parent_trie_input.state);
+                        let nodes_mut = Arc::make_mut(&mut parent_trie_input.nodes);
+                        for block in blocks_iter {
+                            let data = block.trie_data();
+                            state_mut.extend_ref(data.hashed_state.as_ref());
+                            nodes_mut.extend_ref(data.trie_updates.as_ref());
+                        }
+                    }
+                }
+
+                let prefix_sets = hashed_state_for_trie.construct_prefix_sets();
+                let sorted_hashed_state = Arc::new(hashed_state_for_trie.into_sorted());
+                let sorted_trie_updates = Arc::new(trie_output_for_trie.into_sorted());
+
+                parent_trie_input.prefix_sets.extend(prefix_sets);
```

Comment:
The TrieInput for an ExecutedBlock shouldn't have a prefix set, it's never used, so no need to calculate it.